### PR TITLE
Policy as Code for Python

### PR DESCRIPTION
**Pull request by spara**

### CODE REVIEW

File: content/blog/python-for-pac/index.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+PolicyPack(
+    name="kubernetes-python",
+    enforcement_level=EnforcementLevel.MANDATORY,
+    policies=[
+        no_public_services,
+    ],
+)
+```
+
+## Summary
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:181]: Should we add a couple sentences about how to run such a policy pack locally, before the summary? Something like:

> If you'd like to try out one of the examples above, you can create a policy with `pulumi policy new` and to run locally against a Pulumi program specify `pulumi up --policy-pack <path-to-policy-pack>`?

Also wondering if we should mention that you can enforce policy packs in your organization by publishing the policy pack (running `pulumi policy publish [org-name]` inside the policy pack dir). That's what's new in GA, the ability to publish and enforce. Though, that's limited to customers that have support for this enabled in the Pulumi Console. So I could also see leaving it off.

File: content/blog/python-for-pac/index.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+Policy as Code is an important tool for building secure and efficient cloud infrastructure. Pulumi lets you efficiently test resources before deployment and entire stacks when deployed. Learn more about what you can do with Policy as Code:
+
+- [Get Started with Policy as Code]({{< relref "/docs/get-started/crossguard">}})
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:185]: Nit:

```suggestion
- [Get Started with Policy as Code]({{< relref "/docs/get-started/crossguard" >}})
```

File: content/blog/python-for-pac/index.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    policies=[
+        storage_bucket_no_public_read,
+    ],
+)
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+In the case of this Kubernetes policy example, the policy does not allow deploying a loadbalancer which would open the kubernetes cluster to the public Internet.
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:146]: Ah, doesn't look like there's a way to use "k8s" or "kubernetes" with `{{< chooser cloud "aws,azure,gcp" >}}`. @cnunciato, is that by-design?

Comment (cnunciato) [reply content/blog/python-for-pac/index.md:146]: @justinvp Totally an oversight! You should be able to make this work by adding a few lines to `chooser.tsx`:

```diff
 --- a/components/src/components/chooser/chooser.tsx
+++ b/components/src/components/chooser/chooser.tsx
@@ -6,7 +6,7 @@ import { setLanguage, setK8sLanguage, setOS, setCloud } from "../../store/action
 export type LanguageKey = "javascript" | "typescript" | "python" | "go" | "csharp" | "fsharp" | "visualbasic"
 export type K8sLanguageKey = "typescript" | "yaml" | "typescript-kx"
 export type OSKey = "macos" | "linux" | "windows"
-export type CloudKey = "aws" | "azure" | "gcp"
+export type CloudKey = "aws" | "azure" | "gcp" | "kubernetes";

 export type ChooserMode = "local" | "global";
 export type ChooserOptionStyle = "tabbed" | "none";
@@ -408,5 +408,10 @@ export class Chooser {
             name: "GCP",
             preview: false,
         },
+        {
+            key: "kubernetes",
+            name: "Kubernetes",
+            preview: false,
+        },
     ];
 }```

File: content/blog/python-for-pac/index.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
+---
+title: "Policy as Code with Python"
+date: 2020-06-16
+meta_desc: "Policy as Code for Python available as a preview in the 2.0 release."
```
[the remark this answers is not in the document]
Comment (spara) [reply content/blog/python-for-pac/index.md:4]: I think it's better to announce it's GA

File: content/blog/python-for-pac/index.md (hunk 3 above)
Comment (spara) [reply content/blog/python-for-pac/index.md:146]: added k8s to chooser

File: content/blog/python-for-pac/index.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+Devops benefits greatly from policy as code because policies can use software development practices and verify resources without deployments that are costly in time and money. Organizations can also gain significant benefits through cost savings, fine-grained control over infrastructure, efficient deployments, improved compliance, and better use of cloud provider native resources.
+
+## How Policy as Code Works
+
+Policies enforce specific criteria for a resource or for a set of resources (stacks)). For example, a common policy is to ensure that storage is not publicly accessible over the Internet or to ensure that virtual machines must have a firewall.
+
+Pulumi supports two types of polices:
+
+- **ResourceValidationPolicy** validates individual resources in a stack and is run during preview.
```
[the remark this answers is not in the document]
Comment (spara) [reply content/blog/python-for-pac/index.md:25]: Yes, I've included an edited version

File: content/blog/python-for-pac/index.md (hunk 1 above)
Comment (spara) [reply content/blog/python-for-pac/index.md:181]: I'll add the suggestion to try out the policy but leave out publishing in an org

File: content/blog/python-for-pac/index.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+Devops benefits greatly from policy as code because policies can use software development practices and verify resources without deployments that are costly in time and money. Organizations can also gain significant benefits through cost savings, fine-grained control over infrastructure, efficient deployments, improved compliance, and better use of cloud provider native resources.
+
+## How Policy as Code Works
+
+Policies enforce specific criteria for a resource or a set of resources (stacks). For example, a common policy is to ensure that storage is not publicly accessible over the Internet or that virtual machines must have a firewall. Policies are enforced as either *advisory*, which prints a warning message the resource violates the policy; or
+*mandatory*, which prevents a resource deployment if it violates the policy.
+
+Pulumi supports two types of policies:
+
+- **ResourceValidationPolicy** validates individual resources in a stack and is run during preview.
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:26]:
```suggestion
- **ResourceValidationPolicy** validates inputs of individual resources in a stack before the resource is created or modified.
```

File: content/blog/python-for-pac/index.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+- **StackValidationPolicy** validates the stack as a whole and runs when the stack is deployed because some resources may need to exists to validate inputs or outputs.
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:27]:
```suggestion
- **StackValidationPolicy** validates outputs of all resources in the stack after all resources have been created or modified.
```

File: content/blog/python-for-pac/index.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+ResourceValidationPolicy and StackValidationPolicy both run during previews and updates.
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:29]:
```suggestion
ResourceValidationPolicy and StackValidationPolicy validations run during previews and updates. During previews, resources aren't created or modified, so the ResourceValidationPolicy and StackValidationPolicy validations see a preview of what's going to happen (as best we can determine).
```

File: content/blog/python-for-pac/index.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+
+ResourceValidationPolicy validates inputs of individual resources before the resource is created or modified. StackValidationPolicy validates outputs of all resources in the stack after all resources have been created or modified. During previews, resources aren't created or modified, so the ResourceValidationPolicy and StackValidationPolicy validations see a preview of what's going to happen (as best we can determine)
+
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:32]:
```suggestion

```

File: content/blog/python-for-pac/index.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+With stack validation policies, the validation happens after resources have been created or modified. Modifications to resources can't be prevented, but using mandatory enforcement results in a deployment failing (despite the modified resources). If you run Pulumi preview before an update and use a mandatory StackValidationPolicy, you can catch the problems before a real deployment occurs.
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:33]:
```suggestion
With stack validation policies, the validation happens after resources have been created or modified. Modifications to resources can't be prevented, but using mandatory enforcement results in a deployment failing (despite the modified resources). If you run `pulumi preview` before an update and use a mandatory StackValidationPolicy, you can catch the problems before a real deployment occurs.
```

File: content/blog/python-for-pac/index.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+| When is the check performed?   | Before resources are created/modified | After all stack resources have been created/modified                                 |
+| What information is available? | Resource _input_ properties           | Resource _output_ properties (Note: inputs are propagated to outputs during preview) |
+
+Now that we have the basics of policy as code let's take a look at example policies written in Python.
+
+## Examples
+
+Policies are validation functions that validate resources in a Pulumi stack. In the examples below, we define a function that takes `ResourceValidationArgs` and `ReportViolation` as arguments. The function checks to see what type of resource and its arguments, these are used to determine if the resource violates the policy. If the resource is not compliant, the function reports the violation, and because the enforcement level is *mandatory*, the resource will not be deployed.
+
+In the case of this Kubernetes policy example, the policy does not allow deploying a load balancer, which would open the kubernetes cluster to the public Internet.
```
Comment (justinvp) [inline content/blog/python-for-pac/index.md:49]: Is this sentence needed now that k8s is in the chooser? None of the other clouds have similar sentences.

```suggestion
```

**Pull request merged**